### ISSUE TITLE
racy use of shared file

### ISSUE THREAD

**Issue by chlunde** [coreos/console-login-helper-messages#35]
On RHCOS nodes we sometimes see this systemd unit in a failed state:

```
 systemd[1]: Starting Generate /run/issue.d/console-login-helper-messages.issue...
 issuegen[984000]: cat: /run/console-login-helper-messages/40_console-login-helper-messages.issue.staged: No such file or directory
 systemd[1]: console-login-helper-messages-issuegen.service: Main process exited, code=exited, status=1/FAILURE
 systemd[1]: console-login-helper-messages-issuegen.service: Failed with result 'exit-code'.
 systemd[1]: Failed to start Generate /run/issue.d/console-login-helper-messages.issue.
 systemd[1]: console-login-helper-messages-issuegen.service: Consumed 20ms CPU time
```

This might be due to a race between two invocations, one from `/usr/lib/udev/rules.d/90-console-login-helper-messages-issuegen.rules` and one from `/usr/lib/systemd/system/console-login-helper-messages-issuegen.service`, but I do not see any logs from udev so I cannot know for sure.

# Reproducer

```bash
sudo bash -c 'while systemctl restart console-login-helper-messages-issuegen.service; do systemctl reset-failed console-login-helper-messages-issuegen.service; done'
```

```bash
sudo bash -c 'while /usr/libexec/console-login-helper-messages/issuegen remove f; do true; done'
```

**Comment by chlunde** [comment]
Could this possibly be related? I've see this failing together with `systemd-udev-settle.service` several times:

```
Red Hat Enterprise Linux CoreOS 43.81.202002032142.0
WARNING: Direct SSH access to machines is not recommended.

---
Last login: ...
[systemd]
Failed Units: 2
  console-login-helper-messages-issuegen.service
  systemd-udev-settle.service
```

See also https://bugzilla.redhat.com/show_bug.cgi?id=1739757

**Comment by chlunde** [comment]
I've checked the logs, and it seems when this unit fails "udevadm settle" already failed a few seconds before. This unit fails *after* udev fails. So they are somehow related, I do not know if this is an additional failure or if this triggers the race on the temp file.

ts | host | message
-- | -- | --
2020-04-02 20:13:00 | compute02-1 | Failed to start Generate /run/issue.d/console-login-helper-messages.issue. 
2020-04-02 20:12:58 | compute02-1 | Failed to start udev Wait for Complete Device Initialization. 
2020-03-31 07:22:09 | control-plane-0 | Failed to start Generate /run/issue.d/console-login-helper-messages.issue. 
2020-03-31 07:22:05 | control-plane-0 | Failed to start udev Wait for Complete Device Initialization.  
2020-03-25 01:30:25 | compute-4 | Failed to start udev Wait for Complete Device Initialization. 
2020-03-25 01:26:32 | compute-1 | Failed to start Generate /run/issue.d/console-login-helper-messages.issue. 
2020-03-25 01:26:27 | compute-1 | Failed to start udev Wait for Complete Device Initialization. 
2020-03-25 01:22:18 | compute-7 | Failed to start udev Wait for Complete Device Initialization. 
2020-03-25 01:17:40 | infra-1 | Failed to start udev Wait for Complete Device Initialization. 
2020-03-25 01:13:52 | infra-0 | Failed to start udev Wait for Complete Device Initialization.

**Comment by rfairley** [comment]
Thank you for reporting!

The problem looks like the race between the two invocations you mentioned - the [part that writes to and deletes the `.staged` file](https://github.com/rfairley/console-login-helper-messages/blob/a75a52342843e43cc3e38bd5455ad6826d9c11ce/usr/libexec/console-login-helper-messages/issuegen#L52-L64) is executed every time the udev rule invokes `issuegen`. It is possible that the udev invocation and `console-login-helper-messages.service`, which is run on boot, execute at the same time and one of the two deletes the file before the other executes the `cat` command to write the `staged` file to the `generated` file.

There is in fact no need to have the `staged` file - I shouldn't have added it in the first place in https://github.com/rfairley/console-login-helper-messages/commit/d8ef47ace29d865d6874f05b7a93b12c59df23ce. I have opened #36 to fix this, with more explanation there. This should fix the error causing `console-login-helper-messages-issuegen.service` to fail.

> I've checked the logs, and it seems when this unit fails "udevadm settle" already failed a few seconds before. This unit fails after udev fails. So they are somehow related, I do not know if this is an additional failure or if this triggers the race on the temp file.

Could there be implicit ordering between `systemd-udev-settle.service` and `console-login-helper-messages-issuegen.service` that would make the likelihood of the race occurring higher if "udev settle" failed? E.g. `console-login-helper-messages-issuegen.service` [waits for `network-online.target`](https://github.com/rfairley/console-login-helper-messages/blob/00b17df7f11b361a86eecd34c33a1be2bfe4e2ae/usr/lib/systemd/system/console-login-helper-messages-issuegen.service#L4-L5). Thinking if the "udev settle" service failed for an unrelated reason after invoking `/usr/lib/udev/rules.d/90-console-login-helper-messages-issuegen.rules`, but terminated earlier than usual, `console-login-helper-messages-issuegen.service` could be started while the rules invokation is still executing. I can try to verify this by checking the unit ordering - however, the immediate issue should be fixed now by #36.

Given the timeline and the `Generate /run/issue.d/console-login-helper-messages.issue` message, it does seem to be only `console-login-helper-messages-issuegen.service` failing from a race triggered by the udev failure condition seen so far. 

Will do another release once the fix is in and will build the RPM for RHCOS.
